[translator/jaeger] Parse and set Jaeger status as the OpenTelemetry specification defines

MrAlias opened the pull request:
The OpenTelemetry specification defines [the OpenTelemetry span status communicated with Jaeger data types as an `otel.status_code` tag having a string value of `ERROR` or `OK`](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/trace/sdk_exporters/non-otlp.md#span-status). Currently, this is incorrectly parsed and transmitted as an integer representation of the pdata span status value. This changes that to comply with the OpenTelemetry specification.

Additionally, the error tag for Jaeger spans takes precedence over the OpenTelemetry status value in defining if a span is errored. This changes the Jaeger proto to traces transformation to not overwrite the status with the OpenTelemetry status.

Resolves #6675

MrAlias committed:
Parse/set Jaeger status as OTel spec defines

The OpenTelemetry specification defines the OpenTelemetry span status is
communicated with Jaeger types as tags having string values of ERROR or
OK. Currently, this is transmitted as an integer representation of the
pdata span status value. This changes that to comply with the
OpenTelemetry specification.

The error tag for Jaeger spans takes precedence over the OpenTelemetry
status value in defining if a span is errored. This changes the Jaeger
proto to traces transformation to not overwrite the status with the
OpenTelemetry status.

MrAlias committed:
Add changes to changelog

bogdandrutu commented on pkg/translator/jaeger/constants.go:
```diff
@@ -18,7 +18,24 @@ import (
 )
 
+const (
+	// Integer types use two's complement arithmetic.
+	maxUint = ^uint(0)
+	maxInt  = int64(maxUint >> 1)
+	minInt  = -maxInt - 1
```
Why not using "math" package? https://pkg.go.dev/math#pkg-constants `math.MaxInt`?

bogdandrutu commented on pkg/translator/jaeger/jaegerproto_to_traces.go:
```diff
@@ -252,30 +270,55 @@ func setInternalSpanStatus(attrs pdata.AttributeMap, dest pdata.SpanStatus) {
+	return "", false
+}
+
+// Allow tests to check int range errors.
+var validIntRange = func(val int64) error {
+	if val > maxInt || val < minInt {
+		return fmt.Errorf("%w: %d", errRange, val)
+	}
+	return nil
+}
```
Where does this requirement with "int" comes from? It is extremely bad design to validate based on the arch type (32/64). It means that if you are running a 32bit version of the collector for the same received data you may get errors that you don't get on a 64bit version of the collector.

MrAlias replied:
Sure, I'll revert to just validating int32. :+1:

jpkrohling reviewed (commented):
I'm also confused by the math, once that's clarified, I'll run a manual e2e test.

MrAlias committed:
Revert to just check code fits int32

bogdandrutu commented on pkg/translator/jaeger/jaegerproto_to_traces.go:
```diff
@@ -252,30 +271,47 @@ func setInternalSpanStatus(attrs pdata.AttributeMap, dest pdata.SpanStatus) {
 	case pdata.AttributeValueTypeInt:
-		codeVal = attrVal.IntVal()
+		i := attrVal.IntVal()
+		if i > math.MaxInt32 || i < math.MinInt32 {
+			return 0, fmt.Errorf("%w: %d", errRange, val)
+		}
```
Do we really care about this check? What value does it bring? Does all this complication comes from the fact that `tracetranslator.StatusCodeFromHTTP` accepts int?

MrAlias replied:
AFAIK, yes. It was there before and I didn't want to introduce an unreported truncation error by just casting to an `int`.

MrAlias replied:
From async conversation: let's update `tracetranslator.StatusCodeFromHTTP` to accept an `int64` in a separate PR an avoid this entirely.

MrAlias replied:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6699

MrAlias committed:
Update sapmreceiver to not rely on status bug

The sapmreceiver tested for an invalid OTel status code to be passed
through to the underlying span status. This behavior has been fixed and
the underlying span status is no longer set to an invalid value,
therefore the sapmreceiver have also been updated to expect this
corrected behavior.

dmitryax commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@
 
 ## 🧰 Bug fixes 🧰
 
+- `jaeger` receiver/exporter: Parse/set Jaeger status as OTel spec defines (#6682)
```
I agree it's more like a bug fix, but can be a breaking change for those who rely on the OpenCensus style status codes. Should we move it to the _Breaking changes_ section?

MrAlias replied:
sure thing

MrAlias committed:
Update changes in changelog as breaking changes

bogdandrutu commented:
Please rebase

MrAlias committed:
Merge remote-tracking branch 'upstream/main' into jaeger-status-conv

MrAlias committed:
Remove unused constant

MrAlias commented:
> Please rebase

@bogdandrutu PTAL

jpkrohling reviewed (commented):
I did a manual test and it looks good. The test included the Jaeger receiver, the Jaeger exporter, and the logging exporter, and I verified that applications using the Jaeger convention (such as the Hot ROD demo) got converted to the internal value and back, both by checking the output of the logging exporter (jaeger -> internal -> OTLP) and Jaeger UI (jaeger -> internal -> jaeger).

```
Span #19
    Trace ID       : 00000000000000005f11a21978859641
    Parent ID      : 61618f8752981d93
    ID             : 398d42739cbddbb3
    Name           : GetDriver
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2021-12-14 11:01:05.999406 +0000 UTC
    End time       : 2021-12-14 11:01:06.026823 +0000 UTC
    Status code    : STATUS_CODE_ERROR
    Status message : 
```

![image](https://user-images.githubusercontent.com/13387/145986568-3dad143a-b2a4-4dc0-a6de-f0546f0d6827.png)

That all said, the only thing preventing me from approving is the empty file under "results". What's its role? Should it be there?

jpkrohling commented:
Given that @bogdandrutu has left a few comments as well, I'll wait for his feedback until tomorrow before merging.
